overlord/state.go

```diff
@@ -20,25 +20,49 @@
 package overlord
 
 import (
+	"encoding/json"
+	"fmt"
 	"io"
 )
 
 // State represents a snapshot of the system state.
-type State struct{}
+type State struct {
+	entries map[string]json.RawMessage
+}
+
+// NewState returns an empty system state.
+func NewState() *State {
+	return &State{
+		entries: make(map[string]json.RawMessage),
+	}
+}
 
 // Get unmarshals the stored value associated with the provided key
 // into the value parameter.
 func (s *State) Get(key string, value interface{}) {
+	err := json.Unmarshal(s.entries[key], value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not retrieve and unmarshal state entry %q: %v", key, err))
+	}
 }
 
 // Set associates value with key for future consulting by managers.
 // The provided value must properly marshal and unmarshal with encoding/json.
 func (s *State) Set(key string, value interface{}) {
+	serialized, err := json.Marshal(value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not marshal value for state entry %q: %v", key, err))
```

Comment:
This seems fine as a panic, though. If the thing in its present version cannot be marshaled as JSON, we have a larger problem, and definitely some missing tests.